src/gConsent/query/query.ts

```diff
@@ -265,3 +266,45 @@ export async function query(
   }
   return toCredentialResult(response);
 }
+
+/**
+ * Query for Access Credential (Access Requests, Access Grants or Access Denials) based on a given filter,
+ * and traverses all of the result pages.
+ *
+ * @param filter The query filter
+ * @param options Query options
+ * @returns an async iterator going through the result pages
+ * @since unreleased
+ *
+ * @example
+ * ```
+ *  const pages = paginatedQuery(
+ *     {},
+ *     {
+ *       fetch: session.fetch,
+ *       queryEndpoint: new URL("https://vc.example.org/query"),
+ *     },
+ *   );
+ *   for await (const page of pages) {
+ *     // do something with the result page.
+ *   }
+ * ```
+ */
+export async function* paginatedQuery(
+  filter: CredentialFilter,
+  options: {
+    fetch: typeof fetch;
+    queryEndpoint: URL;
+  },
+) {
+  let page = await query(filter, options);
+  while (page.next !== undefined) {
+    yield page;
+    // This is a generator, so we don't want to go through
+    // all the pages at once with a Promise.all approach.
+    // eslint-disable-next-line no-await-in-loop
+    page = await query(page.next, options);
+  }
+  // Return the last page.
+  yield page;
```

Comment:
You previously had return here and I looked up why. It not only yields the last page but marks the iterator as done so I thought you were right. Why do you think it should be yield?

Reply:
That's what I initially expected too, and that's why I had a return, but in tests that didn't work as I expected. If I understood correctly, the iterator has a `return` function to mark it done, but here the function is a generator, which returns said iterator, hence the difference. None of the generator examples in the MDN docs use return, and here is an example in the Node CLI:
```
> function* test() {
... yield 1;
... yield 2;
... yield 3;
... return 4;
... }
undefined
> const i = test()
undefined
> Array.from(i)
[ 1, 2, 3 ]
```